sway/tree/container.c

```diff
@@ -109,10 +110,190 @@ static struct sway_container *_container_destroy(struct sway_container *cont) {
+			while (output->children->length) {
+				struct sway_container *child = output->children->items[0];
+				container_remove_child(child);
+				container_add_child(root_container.children->items[p], child);
+			}
+			container_sort_workspaces(root_container.children->items[p]);
+			arrange_windows(root_container.children->items[p],
+				-1, -1);
+		}
+	}
+
+	wl_list_remove(&output->sway_output->destroy.link);
+	wl_list_remove(&output->sway_output->mode.link);
+	wl_list_remove(&output->sway_output->transform.link);
+	wl_list_remove(&output->sway_output->scale.link);
+
+	wl_list_remove(&output->sway_output->damage_destroy.link);
+	wl_list_remove(&output->sway_output->damage_frame.link);
+
+	wlr_log(L_DEBUG, "OUTPUT: Destroying output '%s'", output->name);
+	container_finish(output);
+	return &root_container;
+}
+
+static struct sway_container *container_workspace_destroy(
+		struct sway_container *workspace) {
+	if (!sway_assert(workspace, "cannot destroy null workspace")) {
+		return NULL;
+	}
+
+	// Do not destroy this if it's the last workspace on this output
+	struct sway_container *output = container_parent(workspace, C_OUTPUT);
+	if (output && output->children->length == 1) {
+		return NULL;
+	}
+
+	struct sway_container *parent = workspace->parent;
+	if (workspace->children->length == 0) {
+		// destroy the WS if there are no children (TODO check for floating)
+		wlr_log(L_DEBUG, "destroying workspace '%s'", workspace->name);
+		ipc_event_workspace(workspace, NULL, "empty");
+	} else {
+		// Move children to a different workspace on this output
+		struct sway_container *new_workspace = NULL;
+		// TODO move floating
+		for (int i = 0; i < output->children->length; i++) {
+			if (output->children->items[i] != workspace) {
+				new_workspace = output->children->items[i];
+				break;
+			}
+		}
+
+		wlr_log(L_DEBUG, "moving children to different workspace '%s' -> '%s'",
+			workspace->name, new_workspace->name);
+		for (int i = 0; i < workspace->children->length; i++) {
+			container_move_to(workspace->children->items[i], new_workspace);
+		}
+	}
+
+	container_finish(workspace);
+	return parent;
+}
+
+
+static void reap_empty_func(struct sway_container *con, void *data) {
+	switch (con->type) {
+		case C_TYPES:
+		case C_ROOT:
+		case C_OUTPUT:
+			// dont reap these
+			break;
+		case C_WORKSPACE:
+			if (!workspace_is_visible(con) && con->children->length == 0) {
+				container_workspace_destroy(con);
+			}
+			break;
+		case C_CONTAINER:
+			if (con->children->length == 0) {
+				container_finish(con);
+			} else if (con->children->length == 1) {
+				struct sway_container *only_child = con->children->items[0];
+				if (only_child->type == C_CONTAINER) {
+					container_remove_child(only_child);
+					container_replace_child(con, only_child);
+					container_finish(con);
+				}
+			}
+		case C_VIEW:
+			break;
+	}
+}
+
+struct sway_container *container_reap_empty(struct sway_container *container) {
```

Comment:
The behavior of this function has changed - it should _always_ be called on the root container otherwise it won't work.